### DISCUSSION TITLE
Need help with using RAG for choosing dynamic tools with ToolProvider

### DISCUSSION THREAD

**Discussion by rwankar** [langchain4j/langchain4j#1915] [Q&A]
With about 25 tools and more to come I'm seeing that the LLMs get confused with the tool choices. The ToolProvider option will do the job except I can't really look for certain words in the user's query to decide which tools to provide. It'll need to be based more on the context as multiple tools will be required with a multi-turn tool calling.

So I thought to use RAG to help choose the right tools. I've tried 2 approaches. I'm not even sure they are different (or correctly implemented).

1. Using a Document object. Here we prepare a Document object from the Tool description and ingest those documents into the EmbeddingStore. Then use the content retriever to match Documents to the user's query.

```

        EmbeddingStoreIngestor toolIngestor = EmbeddingStoreIngestor.builder()
            .documentSplitter(DocumentSplitters.recursive(300, 0))
            .embeddingModel(embeddingModel)
            .embeddingStore(toolEmbeddingStore) // a separate store for toolEmbedding
            .build();
        
        List<Document> documents = new ArrayList<>();
        
        for (ToolSpecification spec : allTools.keySet())
        {
            documents.add(Document.from(spec.description(), Metadata.metadata("id", spec.name())));
        }
        
        // Convert tool definitions into documents and add them to the EmbeddingStore
        toolIngestor.ingest(documents);
        
        toolContentRetriever = EmbeddingStoreContentRetriever.builder()
            .embeddingStore(toolEmbeddingStore)
            .embeddingModel(embeddingModel) // using OpenAIEmbeddingModel
            .maxResults(20) // on each interaction we will retrieve the 2 most relevant segments
            .minScore(0.8) // we want to retrieve segments at least somewhat similar to user query
            .build();

        // Retrieve relevant tool definitions
        List<Content> retrievedContents = toolContentRetriever.retrieve(new Query(userPrompt));

       // loop over the list and get the tool 
       String toolName = content.textSegment().metadata().getString("id");

```

3. Without Document. I think this approach is similar to the Document approach since the embed() method on EmbeddingModel creates a TextSegment from it anyway.

```
public class ToolEmbedding {
        String toolName;
        String toolDescription;

        ToolEmbedding(String toolName, String toolDescription) {
            this.toolName = toolName;
            this.toolDescription = toolDescription;
        }

        String toolName() {
            return toolName;
        }
    }

EmbeddingStore<ToolEmbedding> embeddingStore;

// for each tool create an embedding and add it to the store
Embedding embedding = embeddingModel.embed(toolDescription).content();
embeddingStore.add(embedding, new ToolEmbedding(toolName, toolDescription));

Embedding queryEmbedding = embeddingModel.embed(userQuery).content();

EmbeddingSearchRequest searchRequest = EmbeddingSearchRequest.builder()
       .queryEmbedding(queryEmbedding)
       .maxResults(maxTools)
       .minScore(0.7) // Adjust this threshold as needed
       .build();

EmbeddingSearchResult<ToolEmbedding> relevantTools = embeddingStore.search(searchRequest);

 return relevantTools.matches().stream()
     .map(match -> match.embedded().toolName())
     .collect(Collectors.toList());

```

Neither of the above approaches worked. Both returned all the tools no matter the query. In fact approach with using Documents resulted in duplicate tool names. Note that I have tools that are quite different from each other meaning that some tools should definitely have been dropped for some questions.

Here are my questions

- Is the implementation correct?
- Will it help to provide a much longer description to each tool instead of just a single line description?
- In a multi-turn tool calling how can I use all the messages to determine the tools? It won't help to just look at the latest input.

Thanks in advance.

**Comment by LizeRaes** [comment]
Hi @rwankar a couple of thoughts that might put you on a track that works: 
- [here](https://github.com/LizeRaes/shopping-assistant-light) I used state and steps to route the user query to different AiServices that all have access to a subset of the tools (and RAG and memory) relevant for that state only (even if on the frontend I seem to be chatting with one single AiService). In the readme you see how to try out all the versions (also the one with one big AiService that has all the tools and behaves less predictably)
- to force the model to pick the right tool (and in the right order), I sometimes literally mention 'call tool abc()' in either SystemMessage or UserMessage, this helps a lot (watch out when you refactor the tool name to update it also there though!)
- if you have some rules on how to select the right tool, I could imagine you can add a tool functionSelector(), and tell the model to call this functionSelector() first to determine which other tool(s) it has to call. I expect this approach would work, but curious to hear the result if you'd try it!

**Reply by rwankar** [reply]
Hi @LizeRaes 

Thank you for your suggestions.

> I used state and steps to route the user query to different AiServices that all have access to a subset of the tools (and RAG and memory) relevant for that state only (even if on the frontend I seem to be chatting with one single AiService). In the readme you see how to try out all the versions (also the one with one big AiService that has all the tools and behaves less predictably)

The router option may not work for me as I've determined that even in a subset I might have about 25 tools.

> to force the model to pick the right tool (and in the right order), I sometimes literally mention 'call tool abc()' in either SystemMessage or UserMessage, this helps a lot (watch out when you refactor the tool name to update it also there though!)

The tools are generic in nature and multi-turn and interdependent. So I'm not sure this will work either.

> if you have some rules on how to select the right tool, I could imagine you can add a tool functionSelector(), and tell the model to call this functionSelector() first to determine which other tool(s) it has to call. I expect this approach would work, but curious to hear the result if you'd try it!

I've implemented this right now. In fact I think this is the approach I should take since I need the LLM to decide which tools it needs. I don't think the RAG approach will help.

I've run into an issue here. I provided a ToolProvider to the AiService. On the first query from the user, I return a toolChooser tool with all tool names as boolean parameters with their name and descriptions. The LLM called the toolChooser tool correctly with the right tools. So I keep this list in a thread local and I assumed the ToolProvider will be called so I can return the tool definitions. But that didn't happen. On a tool call from the LLM, the ToolProvider is not invoked. So how can I include the tool definitions in the response?

**Reply by rwankar** [reply]
To elaborate, what I've done is the first request sends the tool chooser tool. In the system prompt I've mentioned that there is a toolChooser that can be called to get the definition of any tool that is required.

```
"tools" : [ {
    "type" : "function",
    "function" : {
      "name" : "toolChooser",
      "description" : "This tool returns all available tools. Call this tool to specify which tools should be included to respond correctly to the user query. Each parameter indicates a tool. Pass true if you need the tool or false.",
      "parameters" : {
        "type" : "object",
        "properties" : {
          "toolA" : {
            "type" : "boolean",
            "description" : "Pass true if... "
          },
          "toolB" : {
            "type" : "boolean",
            "description" : "Pass true if..."
          },
          "toolC" : {
            "type" : "boolean",
            "description" : "Pass true if..."
          }
          ...
       }
    }
  }
}]
```

In the ToolExecutor, I've got this..

```
ToolExecutor toolChooserExecutor = (toolExecutionRequest, memoryId) -> {
        JsonUtils.Object args = JsonUtils.toObject(toolExecutionRequest.arguments());
        // How do I respond with the tool definitions?
};

```

**Reply by rwankar** [reply]
> Please be aware that this approach should help to save on tokens, but not necessarily improve the performance of LLM tool-choosing ability. I assume LLMs are trained first and foremost to choose the right tool in a 1-step approach (where all available tools are present in the request with proper descriptions), not so sure about 2-step.

I wrote an AiService called ToolsAgent. So there are 2 agents: ToolsAgent (only supplies which tools will be required) and ChatAgent (This answers the user's question)

For every user request I first invoke the ToolsAgent for the tool names (which I store) and then the ToolProvider for the ChatAgent provides only those tools. Too early to say but I'm already seeing almost 30% saving on the tokens used. The only issue I see here is that the LLM could miss on asking the right tool and that will be a problem. For this I'm exploring passing the last 5 chat messages of the ChatAgent to the ToolsAgent so it can also have a context of the chat.

> You could try to implement this 2-step approach with a [low-level API](https://docs.langchain4j.dev/tutorials/tools#low-level-tool-api). A little bit more boilerplate code, but much more flexibility. If this approach will prove to work well for your use case, we can consider adding htis feature to AI Services. WDYT?

I am using the low-level API but not sure I follow. Can you please elaborate?

**Comment by LizeRaes** [comment]
@rwankar with your setup, could you try this (elaboration of my suggestion with the ToolChooser:
- you provide in one go all your 25+ tools AND an extra tool called functionSelector() or toolChooser()
- you add something like this to the systemMessage of your main AiService (Chat):
"to determine which tool(s) you should call (and in which order), first of all, call functionSelector(). Then call the tools it specifies."
- about the inner working of toolChooser(), depending on how you decide on the tool(s) to run, you can either implement something rule based there, or call another AiService (that has no tools of it's own) that you prompt to return the toolname (and even extra info on variables) that you want, eg. 
```java
@ApplicationScoped
@RegisterAiService
public interface ToolChooserAiService {

    @SystemMessage("""
          You select the correct tool(s) (in the right order) to be invoked from this list:
          [list of your tools]
          given the userMessage @UserMessage (or memory -> there's options to pass the entire memory if you need it)
          
          here is how you decide on the right tool: if a then b, if c then e, ...
          (or even a few-shot-like approach: here are examples of how to choose the right tool:
          UserMessage: xxx
          Response: tool x, toolz
          
          UserMessage: yyy
          Response: tool g
          
          ...
            """)
    String selectTools(@MemoryId int memoryId, @UserMessage String userMessage);

}
```
You can then use that AiService in the tool by injecting or creating it there and return toolChooserAiService.selectTools(query) from your toolChooser() tool.

I expect this approach can cover quite some complexity. Curious to hear if it works.

**Reply by LizeRaes** [reply]
Nice approach and functionality! Overlooked that, I'll play around with it. It seems to cover this use case well, curious to hear if it works for @rwankar

**Reply by rwankar** [reply]
I haven't been successful at using either TextClassifier or Embedding. I'm still open to exploring this option but the matches are quite poor.

For now, I'm using the LLM itself to choose the tools. For this I have two services.. ToolChooserAiService and ChatAiService.

ChatAiService.toolProvider({
  call ToolChooserAiService with user message to decide which tools
  return only those tools
})

ToolChooserAiService.tools({
 - only provide the toolChooser tool.
 - Each parameter indicates a tool name and is a boolean. Since all parameters are compulsory, the LLM forced to send true/false for each tool (see [my message above](https://github.com/langchain4j/langchain4j/discussions/1915#discussioncomment-10959416))
})

Additionally, I manipulate the ChatMemory of ToolChooserAiService.
- before each call I copy last 5 user messages from the ChatAiService memory into ToolChooserAiService so it has a context since the LLM may ask some questions to the user.. e.g
  - Prompt: I'd like to read a book. ToolChooserAiService Response: findBooks
  - ChatAiService Response to user: What kind of books do you like?
  - User Prompt : I like fiction books
ToolChooserAiService can't only answer on basis of last user input.

In a multi-turn decision making I'm facing a problem. The ChatMessageType only indicates USER, AI, SYSTEM and TOOL_EXECUTOR_RESULT types. It will be very helpful to also tell if a request was a TOOL_EXECUTION_REQUEST. Because in many cases it will help to pass the AI requests that are not execution requests (meant for the user) since they will help in choosing the tool list. Passing the AI requests which contain a tool execution request to the ToolChooserAiService results in an error as the LLM then expects a TOOL_EXECUTION_RESULT as well.

**Reply by rwankar** [reply]
> Nice! How good does it work? What about token usage?

So far I'm seeing about 30% lower token usage. But that could be because we have a lot of tools with a lot of parameters and their descriptions.

> You can filter out AiMessages containing TOOL_EXECUTION_REQUEST by checking AiMessage.hasToolExecutionRequests(), if this is what you mean.

the hasToolExecutionRequests() seems to be on the Response. However I'm copying over the chat memory in the toolProvider code.

```
                List<ChatMessage> messages = chatMemory.messages().stream() // chatMemory is the memory of the ChatAiService
                    .filter(m -> {
                        return m.type() == ChatMessageType.USER;
                    }).collect(Collectors.toList());
                int startIndex = Math.max(messages.size() - 10, 0);
                messages = messages.subList(startIndex, messages.size());
                messages.stream().forEach(m -> toolMemory.add(m)); // toolMemory is the memory of the ToolChooserAiService

```

I don't have the response here. It would be great if there was a ChatMessageType.TOOL_EXECUTION_REQUEST or something like ChatMessage.isToolExecutionRequest().

**Reply by rwankar** [reply]
Thanks! That worked.

One other thing. I call `clear()` on the ChatMemory of the AiToolService each time and then added the messages from the ChatToolService to it. This is so that it starts with a clean slate each time. But I notice that the sequence of the messages gets messed. i.e in the log output I see the USER message go before the SYSTEM prompt. I'm not a 100% sure but I've seen some comments elsewhere say that the SYSTEM message should ideally be the first message.

I tried to find clear all messages except the SYSTEM message using..

```
List<ChatMessage> messages = memory.messages();
messages.removeIf(m -> m.type() != ChatMessageType.SYSTEM);
```

But `memory.messages()` seems to give me a new list, not the underlying list. So that didn't work. And I don't have access to the underlying store to call `store.updateMessages()`

**Reply by rwankar** [reply]
I was trying to achieve the following.

Prompt: Suggest a book
  1. ToolService calls toolChooser() tool
  2. We reply back to ToolService tool with "success".
   -- here I called toolMemory.clear()
  4. Invoke ChatService with memory,
    - System prompt
    - Suggest a book (User Message)
    - tools (findBooks)

Chat Response: What genre do you like?

Prompt: SciFi

At this point I want to remove the ToolExecutionRequest and ToolExecutionResult of the toolChooser call from ToolService chat memory.
So the ToolService ChatMemory should only contain
 - original system prompt of ToolService
 - Suggest a book (User message)
 - What genre do you like? (AI message)
 - SciFi (User message)

I assumed that clearMemory() only cleared the list of messages but retained the System prompt which I don't think it does.

**Reply by rwankar** [reply]
I actually don't want to clear() and add all messages including the system message myself. If I had full control on the messages I could easily do that but ChatMemory.message() returns a new list and I don't seem to have a handle to the underlying store to call updateMessage()

**Reply by rwankar** [reply]
I think I got how to do this.

**Reply by rwankar** [reply]
If I wanted to add a USER message to the chat memory, is it too late to add it in the toolProvider lambda?

```
ToolProvider toolProvider = (chatToolProviderRequest) -> {
        Object chatMemoryId = chatToolProviderRequest.chatMemoryId();
        // call AiToolSerivice to get the relevant tools
        // depending on the tools returned, I want to add some USER message to the chat memory
        return new ToolProviderResult(selectedToolMap);
});
```

It appears like that since the message is added to the chat memory but doesn't go on this call. It goes on the next one.

**Reply by rwankar** [reply]
I don't have an issue with choosing the tools. That part is sorted with calling the LLM with a different agent (although I need to try your suggestion above to see if I can avoid the extra "success" message).

I hope code below helps follow my issue.

```
 AiService chatService = AiServices.builder(ChatService.class) // This is the main chat agent
     .chatLanguageModel(...)
     .chatMemoryProvider(...)
     .systemMessageProvider(...)
     .toolProvider((toolProviderRequest) -> {
         Object chatMemoryId = toolProviderRequest.chatMemoryId();
         ChatMemory chatServiceMemory = getMemory(chatMemoryId);
         Object toolMemoryId = getToolMemoryId(...); //fetch a unique id for current user
         ChatMemory toolMemory = getMemory(toolMemoryId);
         toolMemory.clear();
         // append toolService system message and other user/ai messages from chatServiceMemory
         AiService toolChooserService = AiServices.builder(ToolChooserService.class)
             .chatLanguageModel(...)
             .chatMemory(toolMemory)
             .systemMessageProvider(...)
             .toolProvider(...) // provide the toolChooser tool

         toolChooserService.selectTools(memoryId, userPrompt);
         Result<String> result = toolChooserService.selectTools(toolMemoryId, getText(toolProviderRequest.userMessage())); // getText gets plain text

         // *** ALL GOOD TILL HERE ***
         // The toolChooser tool is called correctly and stores the chosen tools
         // Now, depending on the tools selected, I want to add some User messages to the chatServiceMemory.
         // but what I append to chatMemoryService here is not passed to the LLM.
         // It seems this is too late to modify the messages. It will however show up in the next call to the LLM
         if (tools.contains("toolA") {
             chatMemoryService.add(new UserMessage(...));
         }
     });

chatService.chat(chatMemoryId, userPrompt);

```

Note: The getMemory() function is a helper where I cache the same ChatMemory object that I return in the chatMemoryProvider() on my side so I can clear it after an inactivity period.

I have to change above code to calling each service separately.

```
AiService toolChooserService = AiService.builder()...;
toolChooserService.selectTools(memoryId, userPrompt);

// Finish calling the toolChooserService first.
// Change the chatMemory of the chatService here.
// Then call the ChatService.

AiService chatService = AiServices.builder(ChatService.class)...
chatService.chat(memoryId, userPrompt);
```

**Reply by rwankar** [reply]
Actually the second approach doesn't work either since it too adds the Messages before the SystemPrompt message has been added. So I don't see a way to do this maintaining the exact sequence of conversation.

**Reply by rwankar** [reply]
For now, I landed up fudging this by adding a `ContentRetriever` and inserting the extra details as text into the prompt. However, this way I lose the ability to use the "name" feature on `UserMessage` which I can later use to check if it's present in the chat memory (since I'm using the `TokenWindowChatMemory` and the message could be evicted based on the token size).

It will help to have a hook to append extra messages to the chat memory that retains the order of messages.

**Reply by rwankar** [reply]
> It seems so, we prepare the list of messages to send to the LLM a bit earlier. I am not sure it is an "issue" we have to fix though.
> Did you consider calling ToolChooserService not within ChatService.toolProvider(), but before calling ChatService? In this way you should be able to control the list of messages sent to the LLM by ChatService

Yes, I tried this. But that does not work as this is the order of messages I want.

System
User
Additional custom messages

I can only get the additional custom messages before the user message. Also, I have to add the system message myself otherwise the order becomes custom messages, system and then user.

**Reply by bsorrentino** [reply]
This need seems most related to  a workflow that involve non only LLM calls but also actions & decisions.

The [langgraph4j]( https://github.com/bsorrentino/langgraph4j) could be fit in, take a look to [Agent Executor](https://github.com/bsorrentino/langgraph4j/tree/main/samples/agent-executor) implementation, hope this help.

**Reply by rwankar** [reply]
@langchain4j / @bsorrentino 
Thank you for the suggestion. Right now I need to go LIVE and so I'm trying to keep things simple. But will explore langgraph4j as the use case gets more complex.

**Reply by bsorrentino** [reply]
Good to know! I'm advocate of the "Less is More" approach, and when you get a chance to evaluate langgraph4j, you'll find that it is both simple and straightforward.

**Comment by smileyboy2019** [comment]
For example, I need to determine the weather in my city before deciding on the mode of transportation to take.
I first call the tools at my location, then the tools of the city, and then the tools of transportation. How to achieve such a chain call?

**Reply by rwankar** [reply]
I'm doing the same thing. If you define your tools with good descriptions and provide some examples in the system prompt then gpt-4o-mini is quite smart at figuring out the call chain. Make sure to set the temperature to almost 0. You can quickly check this using the OpenAI Assistants interface without writing any code.
